reconcile/utils/state.py

```diff
@@ -416,41 +420,70 @@ def transaction(self, key: str, value: Any) -> "_TransactionContext":
         return _TransactionContext(self, key, value)
 
 
+@dataclass
+class StateObj:
+    """Represents a state object with a key and a value."""
+
+    key: str
+    exists: bool
+    new_value: Any
+    current_value: Any = None
+    _changed: bool = False
+
+    @property
+    def value(self) -> Any:
+        if self._changed:
+            return self.new_value
+        return self.current_value
+
+    @value.setter
+    def value(self, value: Any) -> None:
+        self._changed = True
+        self.new_value = value
+
+
+class AbortStateTransaction(Exception):
+    """Raise to abort a state transaction."""
+
+
 class _TransactionContext:
     """A context manager to set a key in the state if no exception occurs."""
 
-    def __init__(
-        self,
-        state: State,
-        key: str,
-        value: Any,
-    ):
+    def __init__(self, state: State, key: str, value: Any):
         self.state = state
-        self.key = key
-        self.value = value
+        self.state_obj = StateObj(key, new_value=value, exists=False)
 
-    def __enter__(self) -> bool:
-        """Return True if the key exists in the state, False otherwise.
+    def __enter__(self) -> StateObj:
+        """Return a state object.
 
         Cache the previous value to avoid unnecessary updates.
         """
-        self._previous_value = None
         try:
-            self._previous_value = self.state[self.key]
-            return True
+            self.state_obj.current_value = self.state[self.state_obj.key]
+            self.state_obj.exists = True
         except KeyError:
-            return False
+            self.state_obj.exists = False
+        return self.state_obj
 
     def __exit__(
         self,
         exc_type: type[BaseException] | None,
         exc_value: BaseException | None,
         traceback: TracebackType | None,
-    ) -> None:
+    ) -> bool | None:
         if exc_type:
             # if an exception occurred, we don't want to write to the state
-            return
-        if self._previous_value == self.value:
+            if isinstance(exc_value, AbortStateTransaction):
+                # user wants to abort the transaction
+                return True
+            return False
+        if self.state_obj.new_value is None:
+            # if the new value is None, we don't want to write to the state
+            return None
+        if self.state_obj.exists and (
+            self.state_obj.current_value == self.state_obj.new_value
+        ):
```

Comment:
there is `_changed` in `StateObj`, can we make it as a smart checker to only flag to `True` when the value is changed to a different one? Or do it without setter hook, implement as a pure reader.

```python
@dataclass
class StateObj:
    key: str
    exists: bool
    value: Any
    current_value: Any

    @property
    def changed(self) -> bool:
        return self.current_value != self.value
```

when create state_obj, set value same as current_value by default, or use user override ones, maybe need to make it explicit for `None` value handling.

Reply:
Done in 606b17778ffe3db4c84e8d3b5866926d70f5899f

Reply:
A few edge cases:

* `changed` will be `True` if set the same value as `_init_value`, lead to one request with not changed value, expected?
* if value type is `bool`, `_init_value` is `True`, after set `.value = False`, get `.value` will return `True` again.

Do we need to tell the difference between set called or set with different value, can this work? Backup `value` after init, later changes only apply to `value`, compare `value` with `_init_value` to determine if it's changed.

```python
@dataclass
class TransitionStateObj:
    key: str
    value: Any = None
    _init_value: Any = field(init=False)

    def __post_init__(self):
        self._init_value = self.value

    @property
    def changed(self) -> bool:
        return self.value != self._init_value
```

btw there is a typo `TransistionStateObj` -> `TransitionStateObj`/ `TransactionStateObj`

Reply:
the `__post_init__` does the trick, thank you. Changed in f800da41882c7d257e01c096c28dedf0f03c63d9 and 2522ec35274ede422cb0ed81517d7f1c9660f5e5

Reply:
now that `changed` covers value eq check, https://github.com/app-sre/qontract-reconcile/commit/f800da41882c7d257e01c096c28dedf0f03c63d9 shouldn't be needed anymore, but keep it is also ok as that's the usage side of this class.

Reply:
Lets keep it.